packages/next-swc/crates/next-api/src/dynamic_imports.rs

```diff
@@ -0,0 +1,395 @@
+///     / https://github.com/vercel/next.js/pull/56389/files#diff-791951bbe1fa09bcbad9be9173412d0848168f7d658758f11b6e8888a021552c
+///     / https://github.com/vercel/next.js/pull/56389/files#diff-c33f6895801329243dd3f627c69da259bcab95c2c9d12993152842591931ff01R557
+///     )
+/// - When running an application,
+///    - Server reads generated `react-loadable-manifest.json`, sets dynamicImportIds with the mapping of the import ids, and dynamicImports to the actual corresponding chunks.
+///      (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/server/load-components.ts#L119 /
+///       https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/server/render.tsx#L1417C7-L1420)
+///    - Server embeds those into __NEXT_DATA__ and sent to the client. (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/server/render.tsx#L1453)
+///    - When client boots up, pass it to the client preload (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/client/index.tsx#L943)
+///    - Loadable runtime injects preload fn to wait until all the dynamic components are being loaded, this ensures hydration mismatch won't occur
+///      (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/shared/lib/loadable.shared-runtime.tsx#L281)
+pub(crate) async fn collect_next_dynamic_imports(
+    entry: Vc<Box<dyn EcmascriptChunkPlaceable>>,
+) -> Result<IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules>> {
+    // Traverse referenced modules graph, collect all of the dynamic imports:
+    // - Read the Program AST of the Module, this is the origin (A)
+    //  - If there's `dynamic(import(B))`, then B is the module that is being
+    //    imported
+    // Returned import mappings are in the form of
+    // (Module<A>, Vec<(B, Module<B>)>) (where B is the raw import source string,
+    // and Module<B> is the actual resolved Module)
+    let imported_modules_mapping = NonDeterministic::new()
+        .skip_duplicates()
+        .visit([Vc::upcast(entry)], get_referenced_modules)
+        .await
+        .completed()?
+        .into_inner()
+        .into_iter()
+        .map(build_dynamic_imports_map_for_module);
+
+    // Consolidate import mappings into a single indexmap
+    let mut import_mappings: IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules> =
+        IndexMap::new();
+
+    for module_mapping in imported_modules_mapping {
+        if let Some(module_mapping) = &*module_mapping.await? {
+            let (origin_module, dynamic_imports) = &*module_mapping.await?;
+            import_mappings
+                .entry(*origin_module)
+                .or_insert_with(Vec::new)
+                .append(&mut dynamic_imports.clone())
```

Comment:
If we use `.extend(dynamic_imports.iter().cloned())`, I think we can avoid the vec allocation?